src/travertino/declaration.py

```diff
@@ -74,19 +76,174 @@ def __str__(self):
+        self.name = name
+        owner._PROPERTIES[owner].add(name)
+        owner._ALL_PROPERTIES[owner].add(name)
+
+    def __get__(self, obj, objtype=None):
+        if obj is None:
+            return self
+
+        value = getattr(obj, f"_{self.name}", None)
+        return self.initial if value is None else value
+
+    def __set__(self, obj, value):
+        if value is self:
+            # This happens during autogenerated dataclass __init__ when no value is
+            # supplied.
+            return
+
+        if value is None:
+            raise ValueError(
+                "Python `None` cannot be used as a style value; "
+                f"to reset a property, use del `style.{self.name}`"
+            )
+
+        try:
+            value = self.choices.validate(value)
+        except ValueError:
+            raise ValueError(
+                f"Invalid value {value!r} for property {self.name}; "
+                f"Valid values are: {self.choices}"
+            )
+
+        if value != getattr(obj, f"_{self.name}", None):
+            setattr(obj, f"_{self.name}", value)
+            obj.apply(self.name, value)
+
+    def __delete__(self, obj):
+        try:
+            delattr(obj, f"_{self.name}")
+        except AttributeError:
+            pass
+        else:
+            obj.apply(self.name, self.initial)
+
+
+class directional_property:
+    DIRECTIONS = [TOP, RIGHT, BOTTOM, LEFT]
+    ASSIGNMENT_SCHEMES = {
+        #   T  R  B  L
+        1: [0, 0, 0, 0],
+        2: [0, 1, 0, 1],
+        3: [0, 1, 2, 1],
+        4: [0, 1, 2, 3],
+    }
+
+    def __init__(
+        self, name_format, choices=None, initial=None, _create_directions=True
+    ):
+        """Define a property attribute that proxies for top/right/bottom/left alternatives.
+
+        :param name_format: The format from which to generate subproperties. "{}" will
+            be replaced with "_top", etc.
+        :param choices: The available choices.
+        :param initial: The initial value for the property.
+        """
+        if _create_directions and choices is None:
+            raise TypeError(
+                f"{self.__class__.__name__}.__init__() missing 1 required positional "
+                "argument: 'choices'"
+            )
+        self.name_format = name_format
+        self.choices = choices
+        self.initial = initial
+        self._create_directions = _create_directions
+
+    def __set_name__(self, owner, name):
+        self.name = name
+        owner._ALL_PROPERTIES[owner].add(self.name)
+
+        # Dynamically create the actual properties. They still need to be defined as
+        # class attributes in order to be in the dataclass init, but they don't have to
+        # be set to anything.
+        if self._create_directions:
+            for direction in self.DIRECTIONS:
+                prop_name = self.format(direction)
+                prop = validated_property(self.choices, self.initial)
```

Comment:
Are there potential directional properties that will need different choices or initial values in different directions then, I guess? I thought they'd all be the same, but it's been years since I've done much of anything with CSS.

Reply:
You're right that directional properties don't change type or validation between fields in the tuple. I guess my real conflict here is that I'm not wild about the inconsistency - that actually defining the atomic properties is *almost* optional. AFAICT, they're only there to provide a type annotation. 

Thinking about it, it's not even 100% clear to me that the fact the type annotation works in the new syntax is actually a reliable mechanism at all, or is just an accident of PyCharm parsing. If I'm reading this right, the property is being overwritten by class setter via the `directional_property` definition. Working out the interpretation here likely requires you to get deep into the weeds of how types are declared, and data classes are constructed: is the type assigned to the symbol, or the specific instance assigned to the symbol? PyCharm might be working because it's doing a naïve parse of the class and determining the type - but is that *actually* how the type should be annotated? Is `Style.property_left.__annotation__` actually assigned as a result of the declaration? Does it matter if `directional_property` comes before or after the individual atomic declarations? And, if all of this is true today... will it be true in the future if there are any modifications to PyCharm, dataclasses, or type annotations?

Essentially what I think I'm looking for the code to pick a lane 😄 . In the old version, `directional_property` was a dumb wrapper that converted tuple syntax to the atomic underlying properties. This version is kind of halfway - it partially takes over the construction of the atomic properties... but not enough to actually finish the job with types as well. My position is that it should either take over the definition of the child properties entirely (making the child definitions unnecessary), or revert to being an entirely dumb wrapper. Personal preference is the latter, because it's explicit about every property, and doesn't require any understanding of how `directional_property` works to read the declaration of `property_left` - but I'm open to being convinced otherwise. 

(The other benefit is that the "dumb" approach should be less work, AFAICT 😄)

Reply:
> PyCharm might be working because it's doing a naïve parse of the class and determining the type [...]

I normally use Sublime — in fact I downloaded PyCharm just to work on this — so I'm by no means an expert on how IDEs resolve code completion. However, from messing around with it, it does seem that this situation relies on a very simple special-casing of classes declared with the `@dataclass` decorator. It's smart enough to disable it if it sees `@dataclass(init=False)`, but it's not smart enough to, for instance, realize a class is a dataclass if you "manually" make it one, i.e. `SomeClass = dataclass(SomeClass)`. And the trick I used in the test, passing dataclass a conditionally-assigned dict based on Python version, is apparently ignored; PyCharm still assumes there are keyword parameters, even though there may or may not be, and even if there *won't* be under the currently set Python interpreter.

> I guess my real conflict here is that I'm not wild about the inconsistency - that actually defining the atomic properties is almost optional. AFAICT, they're only there to provide a type annotation.

It's true, there only there so a) PyCharm will see them for code completion and b) dataclass will include them in its constructor parameters. Furthermore... the dataclass decorator *itself* is only there for code completion in the first place. It's otherwise completely unnecessary at runtime, which I'll admit, does bother me.

My personal preference, if a lane is being picked (and if it's possible) is to commit to only declaring the overall directional property. For one thing, it avoids copy-and-pasting the same information four and a half times, so it's quicker and easier to both read and update if needed. And conceptually, if we think of the definition of a class like Pack as a description or spec for a style system, it just makes sense to describe all four together as a unit. Top padding and left padding are not fundamentally different concepts, and it makes little sense to describe each individually; they're intrinsically related subunits of a unifying property.

So, here's a crazy idea... I know #11 was inspired in the first place by a desire for code completion. But if `@dataclass` is there *solely* for that, and the fact that it does so both depends on adding otherwise unnecessary definitions and also may or may not be future proof... What if we just leave it off?  It's true that we wouldn't gain the requested feature for IDE users, but we'd still be getting a cleaner syntax for defining style properties. If we did that, there'd be absolutely no reason to keep the individual `thing_top: int; thing_right: int` folderol.

So long story medium, I suppose I'm proposing two options:

1. **Dataclass decorator + individual property declarations.**
   **Pro:** IDE code completion (at least in PyCharm), and individual properties are more explicit.
   **Con:** requires repetition of definitions, requires the dataclass decorator for no runtime benefit, and relies on code completion implementation which may or may not change in future.
2. **No dataclass decorator + only one declaration for a directional property.**
    **Pro:** Subclassing code doesn't need to import and use dataclass, nor define directional properties individually — things are cleaner, simpler, and less repetitive. The descriptor classes can even lose their clause checking if they're being assigned to themselves, and we no longer have to check Python version.
    **Con:** No code completion for constructor.

It took me the course of writing all this to realize that I'm leaning toward <span>#</span>2. But it is definitely a tradeoff. What do you think?

Reply:
> > Essentially what I think I'm looking for the code to pick a lane 😄 .
> 
> My personal preference, if a lane is being picked (and if it's possible) is to commit to only declaring the overall directional property. For one thing, it avoids copy-and-pasting the same information four and a half times, so it's quicker and easier to both read and update if needed. And conceptually, if we think of the definition of a class like Pack as a description or spec for a style system, it just makes sense to describe all four together as a unit. Top padding and left padding are not fundamentally different concepts, and it makes little sense to describe each individually; they're intrinsically related subunits of a unifying property.

I can certainly understand that position, and don't fundamentally object to it. My concerns are largely practical - whether this position can be reconciled with type annotations, code completion, and backwards compatibility.
 
> So, here's a crazy idea... I know #11 was inspired in the first place by a desire for code completion. But if `@dataclass` is there _solely_ for that, and the fact that it does so both depends on adding otherwise unnecessary definitions and also may or may not be future proof... What if we just leave it off? It's true that we wouldn't gain the requested feature for IDE users, but we'd still be getting a cleaner syntax for defining style properties. 

So - for me, literally the only reason to deal with the code churn is for the code completion and typing improvements. We have an implementation that works fine. Is it perfect? No. If we were starting from scratch, would we build it exactly the same way? Probably Not. Will we remove a source of bugs or open new usage possibilities by changing it in the way you describe? Unless I'm missing something... also no.

> So long story medium, I suppose I'm proposing two options:
> 
> 1. **Dataclass decorator + individual property declarations.**
> 2. **No dataclass decorator + only one declaration for a directional property.**
>
> It took me the course of writing all this to realize that I'm leaning toward #2. But it is definitely a tradeoff. What do you think?

Unless we're actually gaining something substantial from making this change, I don't think the juice is worth the squeeze. Addressing the duplication in directional property definitions is, for me, a minor cosmetic gain... but the existing duplication isn't *that* onerous, and hasn't actually been the source of any bugs. Of all the issues we've got with pack, 4 lines of code duplication in the declaration of directional properties isn't one of them :-)

Reply:
I guess so, yes. Glad I could help by writing a long explanatory note but miss the one *important* detail 🤪

Reply:
Heh, no worries! I appreciate the explanation, and your reasons make sense to me. Thanks for reminding me not to let perfect be the enemy of good.

Reply:
Okey-dokey, that was a pretty quick fix. Still passes Toga's tests, too.